dianatanase/assignment2/Q1.java

```diff
@@ -0,0 +1,159 @@
+import java.util.ArrayList;
+
+// Represent the Tree with Nodes that store data of generic type T.
+// Each node as a key, I assumed keys are positive integers and unique for each node.
+// Each node has a leaf and right child (could be null = leaf).
+// Store the parent of each node as well.
+class Node<T>{
+	
+	T data = null;
+	Node<T> left = null;
+	Node<T> right = null;
+	Node<T> parent = null;
+	int key = -1;
+	
+	Node(Node<T> parent, int key, T data, Node<T> left, Node<T> right){
+		this.data = data;
+		this.parent = parent;
+		this.left = left;
+		this.right = right;
+		this.key = key;
+	}
+	
+	// Update left node, right node and the data.
+	public void leftNode(Node<T> left){
+		this.left = left;
+	}
+	
+	public void rightNode(Node<T> right){
+		this.right = right;
+	}
+	
+	public void updateData(T data){
+		this.data = data;
+	}
+	
+	public T getData(){
+		return data;
+	}
+	
+	// Update parent in case we want to move a node. 
+	public void updateParent(Node<T> parent){
+		this.parent = parent;
+	}
+	
+	public Node<T> getLeft(){
+		return left;
+	}
+	
+	public Node<T> getRight(){
+		return right;
+	}
+	
+	public Node<T> getParent(){
+		return parent;
+	}
+	
+	public int getKey(){
+		return key;
+	}
+}
+
+// Create a Tree structure, each tree is represented by its root Node.
+class Tree<T>{
+	Node<T> root = null;
+	
+	Tree(){}
+	
+	Tree(Node<T> root){
+		this.root = root;
+	}
+	
+	public Node<T> getRoot(){
+		return root;
+	}
+	
+	public boolean isEmpty(){
+		return root == null;
+	}
+}
+
+
+public class Q1 {
+
+	// Get the ancestors of a node, return an ArrayList of integers storing the keys of the ancestors.
+	public static <T> ArrayList<Integer> getAncestorsOfNode(Node<T> node){
+		
+		// Return empty ArrayList if node query is null.
+		if(node == null){
+			System.out.println("Invalid query. Node is null.");
+			return new ArrayList<Integer>();
+		}
+		
+		ArrayList<Integer> ancestors = new ArrayList<Integer>();
+		Node<T> tmp_child = node;
+		Node<T> tmp_parent = node.getParent();
+		
+		// Traverse from the node query up through its parents and the ancestors until the root was added.
+		// Root has null parent.
+		while(tmp_parent != null){
+			ancestors.add(tmp_parent.getKey());
+			tmp_child = tmp_parent;
+			tmp_parent = tmp_child.getParent();
+		}
+		
+		return ancestors;
+	}
+	
+	// Search for a node in a tree and return it, given its key.
+	public static <T> Node<T> findNode(int key, Node<T> root){
+		
+		if(root == null){
+			return null;
+		}
+		
+		if(key < 0){ // If we assume only positive integers are used for keys.
+			System.out.println("Invalid query. Key is negative, should be positive.");
+			return null; 
+		}
+		
+		// Check if the root is the wanted node.
+		int tmp_key = root.getKey();
```

Comment:
No need for the variable tmp_key. Just inline root.getKey() into the equality check.